crates/ark/src/plots/graphics_device.rs

```diff
@@ -70,147 +56,272 @@ thread_local! {
+    ///
+    /// Practically this seems okay, it is just something to keep in mind.
+    #[tracing::instrument(level = "trace", skip_all)]
+    fn process_rpc_requests(&self) {
         // Don't try to render a plot if we're currently drawing.
-        if self._mode.get() != 0 {
+        if self.is_drawing.get() {
+            log::trace!("Refusing to render due to `is_drawing`");
             return;
         }
 
-        // Don't try to render a plot if the 'holdflush' flag is set.
-        if self._holdflush.get() > 0 {
+        // Don't try to render a plot if someone is asking us not to, i.e. `dev.hold()`
+        if !self.should_render.get() {
+            log::trace!("Refusing to render due to `should_render`");
             return;
         }
 
-        // Collect existing channels into a vector of tuples.
+        // Collect existing sockets into a vector of tuples.
         // Necessary for handling Select in a clean way.
-        let channels = {
+        let sockets = {
             // Refcell Safety: Clone the hashmap so we don't hold a reference for too long
-            let channels = self._channels.borrow().clone();
-            channels.into_iter().collect::<Vec<_>>()
+            let sockets = self.sockets.borrow().clone();
+            sockets.into_iter().collect::<Vec<_>>()
         };
 
-        // Check for incoming plot render requests.
+        // Dynamically load all incoming channels within the sockets into a single `Select`
         let mut select = Select::new();
-        for (_id, channel) in channels.iter() {
-            select.recv(&channel.incoming_rx);
+        for (_id, sockets) in sockets.iter() {
+            select.recv(&sockets.incoming_rx);
         }
 
-        let selection = unwrap!(select.try_select(), Err(_error) => {
-            // We don't log errors here, since it's most likely that none
-            // of the channels have any messages available.
-            return;
-        });
-
-        let plot_id = unsafe { &channels.get_unchecked(selection.index()).0 };
-        let socket = unsafe { &channels.get_unchecked(selection.index()).1 };
-        let message = unwrap!(selection.recv(&socket.incoming_rx), Err(error) => {
-            log::error!("{}", error);
-            return;
-        });
-
-        // Get the RPC request.
-        if socket.handle_request(message, |req| self.handle_rpc(req, plot_id)) {
-            return;
+        // Check for incoming plot render requests.
+        // Totally possible to have >1 requests pending, especially if we've plotted
+        // multiple things in a single chunk of R code. The `Err` case is likely just
+        // that no channels have any messages, so we don't log in that case.
+        while let Ok(selection) = select.try_select() {
+            let id = unsafe { &sockets.get_unchecked(selection.index()).0 };
+            let socket = unsafe { &sockets.get_unchecked(selection.index()).1 };
```

Comment:
Not that it matters much but I think get().unwrap() or expect() would be better than unsafe here?